README for testing framework (#753)
broadinstitute/warp: tests/README.md

# Testing in WARP

## :book: Table of Contents

- [Testing in WARP](#testing-in-warp)
  - [:book: Table of Contents](#book-table-of-contents)
  - [:dna: Overview](#dna-overview)
    - [Why test?](#why-test)
  - [:page_with_curl: Terms and Concepts](#page_with_curl-terms-and-concepts)
      - [Main Workflow/Pipeline](#main-workflowpipeline)
      - [Wrapper Workflow](#wrapper-workflow)
      - [Validation Workflow](#validation-workflow)
      - [Plumbing vs Scientific data](#plumbing-vs-scientific-data)
      - [Truth Branches](#truth-branches)
      - [Results vs Storage buckets](#results-vs-storage-buckets)
  - [:hammer_and_pick: Testing Process](#hammer_and_pick-testing-process)
    - [Scala Framework](#scala-framework)
      - [Command Format](#command-format)
    - [What does the wrapper workflow do?](#what-does-the-wrapper-workflow-do)
  - [:mag_right: Working with tests locally](#mag_right-working-with-tests-locally)
    - [Creating a new test](#creating-a-new-test)
    - [Running your new test](#running-your-new-test)

## :dna: Overview 

> This README serves as a summary of the testing process for the pipelines that exists in WARP.

Reviewer: ```suggestion
This README serves as a summary of the testing process for the pipelines that exist in WARP.
```

#### Validation Workflow

The validation workflow takes the output of a completed pipeline and compares it against a known truth set for that pipeline. As with the wrapper workflow, each main workflow must also have a [validation workflow](../verification/).

#### Plumbing vs Scientific data

Each main workflow should have two sets of test data, plumbing and scientific. The plumbing test data is typically a set of downsampled inputs designed to test that the pipeline runs to completion without error. The scientific test data is a set of full-sized inputs designed to thoroughly test the pipeline.

These test inputs exist alongside each pipeline in their respective `test_inputs` directory.

#### Truth Branches

The testing framework has a concept of "truth branches". This allows users to define a branch of truth output to run against.

In most cases the branch being run against is "master" which corresponds to a google storage bucket containing the agreed upon output of the pipeline as it exists in the "master" git branch.

#### Results vs Storage buckets

> Every workflow run in cromwell has its output stored in an "execution" bucket. When running the pipeline tests its important that we store this ouptut in a location that we can later reference if needed.

Reviewer 2: ```suggestion
Every workflow run in cromwell has its output stored in an "execution" bucket. When running the pipeline tests it's important that we store this ouptut in a location that we can later reference if needed.
```

All test workflows that are run will have their output copied to a "results" bucket -> `gs://broad-gotc-test-results/`

> When we want to update the "truth" branches for our pipelines we can copy that output to the respective "truth" bucket -> `gs://broad-gotc-test-storage/`

Reviewer: I thought we had an update truth command? Is it a manual copy?

Author: Yeah, this is just referring to what the workflow actually does, let me try to make it a little more clear

## :hammer_and_pick: Testing Process

We use Jenkins to run our test suite with two webhooks to kick off tests: PR to dev branch kicks off plumbing tests, PR to main/master kicks off scientific tests.

Jenkins will run a job called `Smart Tests` which can be viewed in the [dsp-jenkins](https://github.com/broadinstitute/dsp-jenkins/blob/master/jobs/gotc-jenkins/WarpSmartTestJob.groovy) repo.

The job of the smart tests is to run the `get_changed_pipeline_worklow_test_args.sh` bash script. The script performs a git diff against the target branch and recursively searches for any pipelines that have been changed either directly or indirectly via an imported sub-workflow.

> For each pipeline that has changed it will prepare the arguments for the Scala tool that submits the wrapper workflow to cromwell.

Reviewer: ```suggestion
For each pipeline that has changed it will prepare the arguments for the Scala tool that submits the wrapper workflow to Cromwell.
```

Finally, it passes these arguments to the downstream [Workflow Test](https://github.com/broadinstitute/dsp-jenkins/blob/master/jobs/gotc-jenkins/WarpWorkflowTestsJob.groovy) job which calls the Scala `CloudWorkflow` command and submits the wrapper workflow.

### Scala Framework

As mentioned above, the test framework leverages Scala as a CLI tool to submit the wrapper workflows to Cromwell. The code for the Scala tool can be found in the [tests/broad/scala_test](../tests/broad/scala_test/) directory.

> The Scala tool is responsible for preparing the inputs for the wrapper workflow of the given pipeline and submitting that workflow to the cromwell environment specified.

Reviewer: ```suggestion
The Scala tool is responsible for preparing the inputs for the wrapper workflow of the given pipeline and submitting that workflow to the Cromwell environment specified.
```

> **NOTE** - When looking at the the scala code you will see that there are multiple 'tester' files. A major rewrite of the framework was done to only use one tester command (CloudWorkflow) and to pass in the desired pipeline as an argument. This change makes it easier for contributors and maintainers of these workflows to write tests by pushing the functionality of the test framework away from Scala and into the wrapper workflows.

Reviewer: Will these other tester files be cleaned up when we do the 'clean up old test infra' ticket?

Author: Not my call, you guys can decide what you would like to do with it, leave it there for posterity or rip it out. You will run the risk of breaking some things if you decide to remove things!

  -e <value> | --env <value>
        The environment that this should run in [test|staging|dev|prod]
  --update-truth
        Update the truth data with the results of this run.
  -u | --uncached
        Disable call-caching for the main workflow

# Run the arrays plumbing test against the master truth set in the cromwell dev environment
$ CloudWorkflow -p Arrays -e dev -b master -t Plumbing

# Update the master truth branch for the arrays plumbing test in the cromwell dev environment
$ CloudWorkflow -p Arrays -e dev -b master -t Plumbing --update-truth

# Update the develop truth branch for the WGS scientific test in the cromwell staging environment
$ CloudWorkflow -p WholeGenomeGermlineSingleSample -e staging -b develop -t Scientific --update-truth

# Run the WGS scientific test against the develop truth set in the cromwell staging environment
$ CloudWorkflow -p WholeGenomeGermlineSingleSample -e staging -b develop -t Scientific 
```

### What does the wrapper workflow do?

As mentioned previously, each main workflow will have an accompanying wrapper workflow to act as a test harness. The steps that this workflow takes are the following:

1. Execute the main workflow

> 2. Collect the outputs from the main workflow, these are typically seperated into 'regular' outputs and metrics based outputds

Reviewer 2: ```suggestion
2. Collect the outputs from the main workflow, these are typically separated into 'regular' outputs and metrics based outputs
```
